domain/user/service/service.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"github.com/juju/juju/core/user"
+	usererrors "github.com/juju/juju/domain/user/errors"
+)
+
+// State describes retrieval and persistence methods for upgrade info.
+type State interface {
+	// AddUser will add a new user to the database. If the user already exists
+	// an error that satisfies usererrors.AlreadyExists will be returned.
```

Comment:
@anvial, thanks. We should add the AlreadyExists error to this PR and write a test asserting this contract.

Reply:
done